src/coreclr/tools/dotnet-pgo/TypeRefTypeSystem/TypeRefTypeSystemType.cs

```diff
@@ -0,0 +1,252 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Internal.TypeSystem;
+
+namespace Microsoft.Diagnostics.Tools.Pgo.TypeRefTypeSystem
+{
+    class TypeRefTypeSystemType : MetadataType
+    {
+        TypeRefTypeSystemModule _module;
+        List<TypeRefTypeSystemMethod> _methods = new List<TypeRefTypeSystemMethod>();
+        List<TypeRefTypeSystemField> _fields = new List<TypeRefTypeSystemField>();
+        bool? _isValueType;
+        Instantiation? _instantiation;
+        string _name;
+        string _namespace;
+        Dictionary<string, TypeRefTypeSystemType> _nestedType;
+        TypeRefTypeSystemType _containingType;
+
+        public TypeRefTypeSystemType(string nameSpace, string name, TypeRefTypeSystemModule module)
+        {
+            _namespace = nameSpace;
+            _name = name;
+            _module = module;
+        }
+
+        private TypeRefTypeSystemType(string nameSpace, string name, TypeRefTypeSystemType containingType, TypeRefTypeSystemModule module)
+        {
+            _namespace = nameSpace;
+            _name = name;
+            _module = module;
+            _containingType = containingType;
+        }
+
+        public void SetIsValueType(bool isValueType)
+        {
+            if (!_isValueType.HasValue)
+            {
+                _isValueType = isValueType;
+            }
+            else
+            {
+                if (_isValueType.Value != isValueType)
+                {
+                    throw new Exception($"Same type `{ToString()}` used as both ValueType and non-ValueType");
+                }
+            }
+        }
+
+        public void SetGenericParameterCount(int parameterCount)
+        {
+            if (!_instantiation.HasValue)
+            {
+                if (parameterCount == 0)
+                {
+                    _instantiation = Instantiation.Empty;
+                }
+                else
+                {
+                    TypeDesc[] instantiationArgs = new TypeDesc[parameterCount];
+                    for (int i = 0; i < parameterCount; i++)
+                    {
+                        instantiationArgs[i] = new TypeRefTypeSystemGenericParameter(this, i);
+                    }
+                    _instantiation = new Instantiation(instantiationArgs);
+                }
+            }
+            else
+            {
+                if (_instantiation.Value.Length != parameterCount)
+                {
+                    throw new Exception($"Same type `{ToString()}` expected to have both {_instantiation.Value.Length} and {parameterCount} generic arguments");
+                }
+            }
+        }
+
+        public override Instantiation Instantiation
+        {
+            get
+            {
+                if (!_instantiation.HasValue)
+                    SetGenericParameterCount(0);
```

Comment:
Hmm, can you please educate me what this side effect is about?

Reply:
We don't actually have a guarantee that we determine the size of the instantiation while loading types, we only have a guarantee that it will happen if it matters for the output (as the output format from this tool is exceedingly similar to the input format). This logic is in place, so that if we ask the instantiation shape before we find out what it really is. Calling SetGenericParameterCount will force the instantiation count to be constant for the rest of the process, and if we find that it should be something else, it will trigger a fault.

Reply:
Got it, thanks.